haskell/test/SMT.hs

```diff
@@ -0,0 +1,666 @@
+  , ssIsPaused :: Bool
+  , ssTotalSupply :: Natural
+  } deriving stock (Eq, Generic, Show)
+
+instance Buildable SimpleStorage where
+  build = genericF
+
+resultToSs :: InterpretResult -> SimpleStorage
+resultToSs InterpretResult {..} = case cast iurNewStorage of
+  (Just (sval :: T.Value (T.ToT Storage))) -> storageToSs (fromVal @Storage sval)
+  Nothing -> error "Impossible"
+
+storageToSs :: Storage -> SimpleStorage
+storageToSs storage = let
+  StorageMinters ssMintingAllowances = storage
+  StorageLedger ssLedger = storage
+  StorageOperators (T.BigMap ssOperators) = storage
+  StorageRoles (MasterMinterRole ssMasterMinter) = storage
+  StorageRoles (PauserRole ssPauser) = storage
+  StorageRoles (OwnerRole ssOwner) = storage
+  StorageRoles (PendingOwnerRole ssPendingOwner) = storage
+  StorageSafelistContract ssSafelistContract = storage
+  StoragePaused ssIsPaused = storage
+  (_, (_, arg #total_supply -> ssTotalSupply)) = storage
+  in SimpleStorage {..}
+
+ssToOriginationParams
+  :: SimpleStorage
+  -> OriginationParams
+ssToOriginationParams SimpleStorage {..} = let
+  minters = ssMintingAllowances
+  in defaultOriginationParams
+    { opBalances = ssLedger
+    , opOwner = ssOwner
+    , opOwnerToOperators = mkOwnerToOperator ssOperators
+    , opMasterMinter = ssMasterMinter
+    , opPauser = ssPauser
+    , opPaused = ssIsPaused
+    , opMinters = minters
+    , opPendingOwner = ssPendingOwner
+    , opSafelistContract = TAddress <$> ssSafelistContract
+    }
+  where
+    mkOwnerToOperator :: Map (Address, Address) () -> Map Address [Address]
+    mkOwnerToOperator operatorMap =
+      Map.foldrWithKey (\(ow, op) _ m -> Map.alter (alterFn op) ow m) mempty operatorMap
+
+    alterFn :: Address -> Maybe [Address] -> Maybe [Address]
+    alterFn op Nothing = Just [op]
+    alterFn op (Just ops) = Just (op:ops)
+
+data ContractCall p = ContractCall
+  { ccSender :: Address
+  , ccParameter :: p
+  , ccIdx :: Int
+  } deriving stock Generic
+
+instance (Buildable p) => Buildable (ContractCall p) where
+  build = genericF
+
+instance (Buildable p) => Show (ContractCall p) where
+  show = pretty
+
+instance Buildable (Address, Natural) where
+  build  = genericF
+
+instance Buildable (Map Address Natural) where
+  build = pretty . Map.assocs
+
+data ContractState = ContractState
+  { csStorage :: SimpleStorage
+  , csError :: [(Int, ModelError)]
+  } deriving stock (Eq, Generic)
+
+instance Buildable ContractState where
+  build = genericF
+
+instance Show ContractState where
+  show = pretty
+
+-- | This structure holds the state of the generator
+-- based on which new input values are generated.
+data GeneratorState = GeneratorState
+  { gsOwnerPool :: [Address]
+  , gsMasterMinterPool :: [Address]
+  , gsMinterPool :: Map Address Natural
+  , gsPauserPool :: [Address]
+  , gsTokenOwnerPool :: [Address]
+  }
+
+type GeneratorM = ReaderT GeneratorState Gen
+
+instance MonadFail Gen where
+  fail = error "Not implemented"
```

Comment:
It looks fairly hacky to me, can it be avoided? For example, instead of `(owner:_) <- shuffle gsOwnerPool` can you do `owner <- Data.List.head <$> shuffle gsOwnerPool`? Or just use `elements`.

Reply:
Have done this using `elements`.